### PR TITLE
Cluster horizon

### PR DESCRIPTION
**Pull request by cdtn**

### CODE REVIEW

File: seismiqb/src/labels/horizon_visualization.py [hunk 1]
```diff
@@ -78,14 +78,16 @@ def show(self, attributes='depths', mode='imshow', short_title=True, return_figu
             'suptitle_label': f'`{self.name}` on field `{self.field.displayed_name}`',
             **kwargs
         }
-        self.field.show(attributes=attributes, mode=mode, short_title=short_title,
+        figure = self.field.show(attributes=attributes, mode=mode, short_title=short_title,
                         return_figure=return_figure, **kwargs)
```
Comment (HollowPrincess) [inline seismiqb/src/labels/horizon_visualization.py:82]: Identation

File: seismiqb/src/labels/horizon.py [hunk 2]
```diff
@@ -267,7 +267,7 @@ def __sub__(self, other):
         if not isinstance(other, type(self)):
             raise TypeError(f"Operands types do not match. Got {type(self)} and {type(other)}.")
 
-        presence = other.presence_matrix
+        presence = other.full_binary_matrix
```
[the remark this answers is not in the document]
Comment (cdtn) [reply seismiqb/src/labels/horizon.py:270]: IMO this alias introduces more unambiguity since throughout the codebase it's used together with `full_binary_matrix`.

[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply seismiqb/src/labels/horizon.py:270]: We can use `presence_matrix` as an alias to `binary_matrix`, and use `full_presence_matrix` if needed

File: seismiqb/src/labels/horizon_attributes.py [hunk 3]
```diff
@@ -451,14 +435,20 @@ def load_attribute(self, src, location=None, use_cache=True, enlarge=False, **kw
         Load 'metrics' attribute with specific evaluation parameter and following normalization.
         >>> horizon.load_attribute('metrics', metric='local_corrs', normalize='min-max')
         """
-        src = self.ALIAS_TO_ATTRIBUTE.get(src, src)
-        enlarge = enlarge and self.is_carcass
+        kwargs['enlarge'] = enlarge and self.is_carcass
+
+        attr_name = self.ALIAS_TO_ATTR.get(src, src)
+        class_attr = getattr(type(self), attr_name, None)
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:441]: `method_or_property`  variable name would be long, but way easier to comprehend

File: seismiqb/src/labels/horizon_attributes.py [hunk 4]
```diff
@@ -706,3 +696,31 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
             spikes = binary_dilation(spikes, iterations=dilation).astype(np.float32)
             spikes[self.field.zero_traces == 1] = np.nan
         return spikes
+
+    # Clustering
+    @transformable
+    def cluster(self, clusterer, attributes, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:702]: It may sound ironic, but maybe we should move this to a separate PR, and merge everything else first
@cdtn your call

File: seismiqb/src/labels/horizon_attributes.py [hunk 5]
```diff
@@ -164,19 +156,24 @@ def pca_transform(data, n_components=3, **kwargs):
...
+        """ Boolean matrix with `True` values at places where horizon is present and `False` everywhere else. """
+        return (self.matrix != self.FILL_VALUE).astype(np.bool)
 
+    @property
     @lru_cache(maxsize=1)
-    def _presence_matrix(self):
+    def full_binary_matrix(self):
         """ A method for getting binary matrix in cubic coordinates. Allows for introspectable cache. """
-        return self.full_binary_matrix
+        return self.matrix_put_on_full(self.binary_matrix)
```
Comment (HollowPrincess) [inline seismiqb/src/labels/horizon_attributes.py:174]: What is faster:
```return (self.full_matrix != self.FILL_VALUE).astype(np.bool)```
or your implementation?

Comment (cdtn) [reply seismiqb/src/labels/horizon_attributes.py:174]: image

File: seismiqb/src/labels/horizon_attributes.py [hunk 6]
```diff
@@ -163,19 +159,25 @@ def pca_transform(data, n_components=3, **kwargs):
...
+
+    # an alias
+    mask = full_binary_matrix
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:180]: Why is that needed? You added the `mask` to dictionary with aliases

File: seismiqb/src/utils/functions.py [hunk 7]
```diff
@@ -447,3 +447,26 @@ def concat_sorted(first_array, second_array):
 def split_array(array, labels):
     """ Split (groupby) array by values from labels. Labels must be sorted and all groups must be contiguous. """
     return np.split(array, np.unique(labels, return_index=True)[1][1:])
+
+
+@njit
+def get_adjancency_matrix(arr, n):
+    """ Produce an adjacency matrix for a labeled integer-valued array with the provided amount of unique elements. """
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/functions.py:454]: Write a more detailed explanation for this function: usually, `njit`-tted functions are noticeably hard to read after some time. Make it easier for future CDTN:)

File: seismiqb/src/labels/horizon_attributes.py [hunk 8]
```diff
@@ -708,3 +710,82 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
             spikes = binary_dilation(spikes, iterations=dilation).astype(np.float32)
             spikes[self.field.zero_traces == 1] = np.nan
         return spikes
+
+    # Clustering
+    @transformable
+    def cluster(self, clusterer, attributes, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:716]: Should we add default value for `clusterer`?

File: seismiqb/src/labels/horizon_attributes.py [hunk 9]
```diff
@@ -708,3 +710,82 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
...
+        After a certain `threshold` value all labeled objects tend to merge together into a single mega cluster.
+        So keep in mind that while the valid range of values for `threshold` is (0, 1), the truly useful subrange
+        might be much more narrow (0.0, 0.3 for example) depending on the distribution of adjacent objects alikeness.
+        """
+        # re-label clusters splitting disconnected areas into separate objects
+        labeled, num_labeled = label(clustered, background=background, return_num=True)
+
+        # find adjacent objects, drop row and column correponding to background
+        adjacency = get_adjancency_matrix(labeled, num_labeled + 1)
+        adjacency = adjacency.astype(bool)[1:, 1:]
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:769]: You could use `np.bool_` dtype inside the `get_adjacency_matrix` function

File: seismiqb/src/labels/horizon_attributes.py [hunk 10]
```diff
@@ -708,3 +710,82 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
...
+        cluster_labels = clusterer(features, **kwargs)
+
+        cluster_map = np.full(self.field.spatial_shape, fill_value=np.nan, dtype=np.float32)
+        cluster_map[i, x] = cluster_labels
+
+        return cluster_map
+
+    @staticmethod
+    def merge_clusters(clustered, features, threshold=.1, background=-1):
+        """ Analyze clustered horizon map and merge adjacent clusters if averages of their features are similar.
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:744]: Describe the overall idea somewhere: in the beginning of the documentation, for example

File: seismiqb/src/labels/horizon_attributes.py [hunk 11]
```diff
@@ -708,3 +710,82 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
...
+
+        cluster_labels = clusterer(features, **kwargs)
+
+        cluster_map = np.full(self.field.spatial_shape, fill_value=np.nan, dtype=np.float32)
+        cluster_map[i, x] = cluster_labels
+
+        return cluster_map
+
+    @staticmethod
+    def merge_clusters(clustered, features, threshold=.1, background=-1):
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:743]: This method looks beautiful 
🥇

**Review by SergeyTsimfer — APPROVED** [review]

File: seismiqb/src/labels/horizon_attributes.py (hunk 6 above)
Comment (cdtn) [reply seismiqb/src/labels/horizon_attributes.py:180]: So you could request it like this: `horizon.mask`

File: seismiqb/src/labels/horizon_attributes.py (hunk 8 above)
Comment (cdtn) [reply seismiqb/src/labels/horizon_attributes.py:716]: I might add a string value support for this argument, and it will try to import the corresponding class from `sklearn.cluster`. What do you think about it?

File: seismiqb/src/labels/horizon_attributes.py (hunk 11 above)
Comment (cdtn) [reply seismiqb/src/labels/horizon_attributes.py:743]: 😊

File: seismiqb/src/labels/horizon_visualization.py [hunk 12]
```diff
@@ -70,6 +71,7 @@ def _show_add_prefix(attribute, prefix=None):
     def show(self, attributes='depths', mode='imshow', return_figure=False, **kwargs):
         """ Field visualization with custom naming scheme. """
         attributes = DelegatingList(attributes)
+        attributes = attributes.apply(lambda item: copy(item) if isinstance(item, dict) else item)
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_visualization.py:74]: `dict(item)`?

File: seismiqb/src/utils/decorators.py [hunk 13]
```diff
@@ -33,14 +34,19 @@ def transformable(method):
...
+    def wrapper(instance, *args, dtype=None, on_full=False, channels=None, normalize=False, fill_value=None,
+                enlarge=False, enlarge_width=10, atleast_3d=False, n_components=None, **kwargs):
         result = method(instance, *args, **kwargs)
 
         if dtype and hasattr(instance, 'matrix_set_dtype'):
             result = instance.matrix_set_dtype(result, dtype=dtype)
         if on_full and hasattr(instance, 'matrix_put_on_full'):
             result = instance.matrix_put_on_full(result)
+        if channels is not None:
+            if channels == 'mid':
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/decorators.py:46]: `'middle'`?

File: seismiqb/src/labels/horizon_attributes.py [hunk 14]
```diff
@@ -708,3 +710,145 @@ def get_spikes_map(self, spikes_mode='median', threshold=1., dilation=5,
             spikes = binary_dilation(spikes, iterations=dilation).astype(np.float32)
             spikes[self.field.zero_traces == 1] = np.nan
         return spikes
+
+    # Clustering
+    @lru_cache(maxsize=1, apply_by_default=False, copy_on_return=True)
+    @transformable
+    def cluster(self, clusterer, attributes='amplitudes', background=-1, **kwargs):
```
Comment (SergeyTsimfer) [inline seismiqb/src/labels/horizon_attributes.py:717]: Move the entire clustering procedure into a separate `Controller` if `SeismicSolutions`: to reason to merge it in `seismiqb`

File: seismiqb/src/labels/horizon_visualization.py (hunk 12 above)
Comment (cdtn) [reply seismiqb/src/labels/horizon_visualization.py:74]: No, too implicit.
